scipy/signal/_short_time_fft.py

```diff
@@ -0,0 +1,1421 @@
+PAD_TYPE = Literal['zeros', 'edge', 'even', 'odd']
+
+#: Allowed values for property `ShortTimeFFT.fft_typ`:
+FFT_TYP_TYPE = Literal['twosided', 'centered', 'onesided', 'onesided2X']
+
+
+def _calc_dual_canonical_window(win: NDArray, hop: int) -> NDArray:
+    """Calculate canonical dual window for 1d window `win` and a time step
+    of `hop` samples.
+
+    A ``ValueError`` is raised, if the inversion fails.
+
+    This is a separate function not a method, since it is also used in the
+    class method ``ShortTimeFFT.from_dual()``.
+    """
+    if hop > len(win):
+        raise ValueError(f"{hop=} is larger than window length of {len(win)}" +
+                         " => STFT not invertible!")
+    if issubclass(win.dtype.type, np.integer):
+        raise ValueError("Parameter 'win' cannot be of integer type, but " +
+                         f"{win.dtype=} => STFT not invertible!")
+        # The calculation of `relative_resolution` does not work for ints.
+        # Furthermore, `win / DD` casts the integers away, thus an implicit
+        # cast is avoided, which can always cause confusion when using 32-Bit
+        # floats.
+
+    w2 = win.real**2 + win.imag**2  # win*win.conj() does not ensure w2 is real
+    DD = w2.copy()
+    for k_ in range(hop, len(win), hop):
+        DD[k_:] += w2[:-k_]
+        DD[:-k_] += w2[k_:]
+
+    # check DD > 0:
+    relative_resolution = np.finfo(win.dtype).resolution * max(DD)
+    if not np.all(DD >= relative_resolution):
+        raise ValueError("Short-time Fourier Transform not invertible!")
+
+    return win / DD
+
+
+# noinspection PyShadowingNames
+class ShortTimeFFT:
+    r"""Provide a parametrized discrete Short-time Fourier transform (stft)
+    and its inverse (istft).
+
+    The `stft` calculates sequential FFTs by sliding a window (`win`) over
+    an input signal by `hop` increments. It can be used to quantify the change
+    of the spectrum over time.
+
+    The `stft` is represented by a complex-valued matrix S[q,p] where the
+    p-th column represents an FFT with the window centered at the time
+    t[p] = p * `delta_t` = p * `hop` * `T` where `T` is  the sampling interval
+    of the input signal. The q-th row represents the values at the frequency
+    f[q] = q * `delta_f` with `delta_f` = 1 / (`mfft` * `T`) being the bin
+    width of the FFT.
+
+    The inverse STFT `istft` is calculated by reversing the steps of the
+    STFT: Take the IFFT of the p-th slice of S[q,p] and multiply the result
+    with the so-called dual window (see `dual_win`). Shift the result by
+    p * `delta_t` and add the result to previous shifted results to reconstruct
+    the signal. If only the dual window is known and the STFT is invertible,
+    `from_dual` can be used to instanciate this class.
+
+    Due to the convetion of time t = 0 being at the first sample of the input
+    signal, the SFT values typically has negative time slots. Hence,
```

Comment:
```suggestion
    signal, the STFT values typically have negative time slots. Hence,
```

Unless "SFT" stands for something else?